src/utils/pgwire/src/pg_message.rs

```diff
@@ -41,6 +47,82 @@ pub struct FeQueryMessage {
     pub sql_bytes: Bytes,
 }
 
+#[derive(Debug)]
+pub struct FeBindMessage {}
+
+#[derive(Debug)]
+pub struct FeExecuteMessage {
+    pub max_rows: i32,
+}
+
+#[derive(Debug)]
+pub struct FeParseMessage {
+    pub query_string: Bytes,
+}
+
+#[derive(Debug)]
+pub struct FeDescribeMessage {
+    // 'S' to describe a prepared statement; or 'P' to describe a portal.
+    pub kind: u8,
+}
+
+impl FeDescribeMessage {
+    pub fn parse(mut buf: Bytes) -> Result<FeMessage> {
+        let kind = buf.get_u8();
+        let _pstmt_name = read_null_terminated(&mut buf)?;
+
+        if kind != b'S' {
+            unimplemented!("only prepared statement Describe is implemented");
+        }
+
+        Ok(FeMessage::Describe(FeDescribeMessage { kind }))
+    }
+}
+
+impl FeBindMessage {
+    pub fn parse(mut buf: Bytes) -> Result<FeMessage> {
+        let portal_name = read_null_terminated(&mut buf)?;
+        let _pstmt_name = read_null_terminated(&mut buf)?;
+
+        if !portal_name.is_empty() {
+            unimplemented!("named portals not implemented");
+        }
+
+        Ok(FeMessage::Bind(FeBindMessage {}))
+    }
+}
+
+impl FeExecuteMessage {
+    pub fn parse(mut buf: Bytes) -> Result<FeMessage> {
+        let portal_name = read_null_terminated(&mut buf)?;
+        let max_rows = buf.get_i32();
+
+        if !portal_name.is_empty() {
+            unimplemented!("named portals not implemented");
+        }
+
+        if max_rows != 0 {
+            unimplemented!("row limit in Execute message not supported");
+        }
+
+        Ok(FeMessage::Execute(FeExecuteMessage { max_rows }))
+    }
+}
+
+impl FeParseMessage {
+    pub fn parse(mut buf: Bytes) -> Result<FeMessage> {
+        let _pstmt_name = read_null_terminated(&mut buf)?;
+        let query_string = read_null_terminated(&mut buf)?;
+        let nparams = buf.get_i16();
```

Comment:
The pg driver looks strange here. Even if I test with the 'SELECT $1::VARCHAR', &[&'Hello, World']', the nparams is 0 here. cc @ZENOTME

Reply:
I think the nparams is the explicit parameter type number. I debug this procedure and I think the client will use "SELECT $1::TEXT" independently to parse without any explicit parameter type, so the nparams is 0. The parameter(&[&'Hello, World']') means the "real" parameter used to bind to the statement. It has nothing to do with nparams.
By default, the parameter type is inferred by the server, so it generally will be 0 unless you use prepare_type to indicate the explicit parameter type

Reply:
So in this case, we should parse the query and infer the type? 
Can we rewrite the unittest to explicit declare return type? So that it is more tested well.

Reply:
Anyway I think it is ok to merge this one.

Reply:
Before do the sqlparse, we may need move pgwire to `frontend` crates. Otherwise there is a bi-directional dependency.

Reply:
As rust practice, you can try to do some code refactor or clippy fix.